File: Tests/OverlayTests/OverlayCoordinatorTests.swift
import AppKit
import Testing
@testable import Scratchpad

@MainActor
private final class SpyFadeAnimator: OverlayFadeAnimator {
    func fadeIn(_ window: NSWindow, duration: TimeInterval) {}
    func fadeOut(_ window: NSWindow, duration: TimeInterval, completion: @escaping @MainActor () -> Void) {
        completion()
    }
}

@MainActor
private final class StubScreenObserving: ScreenObserving {
    var screens: [NSScreen] { NSScreen.screens }
    func observeScreenChanges(_ handler: @escaping @MainActor () -> Void) -> AnyObject { NSObject() }
}

@MainActor
private func makeCoordinator() -> (OverlayCoordinator, OverlayState, DrawingStoreFactory) {
    let overlayState = OverlayState()
    let drawingStoreFactory = DrawingStoreFactory(repository: FileDrawingRepository(
        directory: FileManager.default.temporaryDirectory.appendingPathComponent("OverlayCoordinatorTests")
    ))
    let coordinator = OverlayCoordinator(
        overlayState: overlayState,
        drawingStoreFactory: drawingStoreFactory,
        renderer: CAShapeLayerStrokeRenderer(),
        animator: SpyFadeAnimator(),
        settingsStore: SettingsStore(persistence: InMemorySettingsPersistenceForOverlayTests()),
        screenObserving: StubScreenObserving()
    )
    return (coordinator, overlayState, drawingStoreFactory)
}

private final class InMemorySettingsPersistenceForOverlayTests: SettingsPersisting {
    func loadSettings() -> AppSettings? { nil }
    func saveSettings(_ settings: AppSettings) {}
}

@MainActor
struct OverlayCoordinatorTests {

    @Test func toggleNeverMutatesAnyDrawingStore() {
        let (coordinator, _, factory) = makeCoordinator()
        guard let screen = NSScreen.main, let display = DisplayIdentifier(screen: screen) else { return }
        let store = factory.store(for: display)
        store.beginStroke(at: .zero)
        store.appendPoint(CGPoint(x: 1, y: 1))
        store.endStroke()
        let strokesBeforeToggle = store.strokes

        coordinator.toggle()
        coordinator.toggle()

        #expect(store.strokes == strokesBeforeToggle)
    }

    @Test func toggleFlipsOverlayVisibility() {
        let (coordinator, overlayState, _) = makeCoordinator()
        #expect(overlayState.isVisible == false)
        coordinator.toggle()
        #expect(overlayState.isVisible == true)
        coordinator.toggle()
        #expect(overlayState.isVisible == false)
    }

    @Test func clearAllNeverMutatesOverlayVisibility() {
        let (coordinator, overlayState, _) = makeCoordinator()
        coordinator.toggle()
        let visibilityBeforeClear = overlayState.isVisible

        coordinator.clearAll()

        #expect(overlayState.isVisible == visibilityBeforeClear)
    }

    @Test func clearAllErasesStrokesOnEveryConnectedDisplay() {
        let (coordinator, _, factory) = makeCoordinator()
        guard let screen = NSScreen.main, let display = DisplayIdentifier(screen: screen) else { return }
        let store = factory.store(for: display)
        store.beginStroke(at: .zero)
        store.appendPoint(CGPoint(x: 1, y: 1))
        store.endStroke()

        coordinator.clearAll()

        #expect(store.strokes.isEmpty)
    }
}

File: Windows/Scratchpad.Tests/OverlayTests/OverlayCoordinatorTests.cs
using System.IO;
using System.Windows;
using System.Windows.Forms;
using Scratchpad.Animations;
using Scratchpad.Drawing;
using Scratchpad.Overlay;
using Scratchpad.Persistence;
using Scratchpad.Preferences;
using Scratchpad.Rendering;
using Scratchpad.Tests.SettingsTests;
using Scratchpad.Utilities;
using Xunit;

namespace Scratchpad.Tests.OverlayTests;

internal sealed class SpyFadeAnimator : IOverlayFadeAnimator
{
    public void FadeIn(Window window, TimeSpan duration) { }
    public void FadeOut(Window window, TimeSpan duration, Action completion) => completion();
}

internal sealed class StubScreenObserving : IScreenObserving
{
    public IReadOnlyList<Screen> Screens => Screen.AllScreens;
    public IDisposable ObserveScreenChanges(Action handler) => new NoopSubscription();

    private sealed class NoopSubscription : IDisposable
    {
        public void Dispose() { }
    }
}

/// These tests construct real `OverlayWindowController`/`OverlayWindow`
/// (WPF `Window`) instances via `Toggle()`/`Show()` -- `SpyFadeAnimator`
/// only skips the animation, mirroring the macOS app's own
/// `OverlayCoordinatorTests`, which does the same with a real `NSPanel`.
/// Never run against a real Windows/WPF test host in this session (no
/// Windows machine was available); if window construction on xUnit's
/// worker thread turns out to need an STA thread with a pumped
/// `Dispatcher`, wrap these in a dedicated STA test thread (e.g.
/// `Xunit.StaFact`) rather than assuming a bare `[Fact]` is sufficient.
public class OverlayCoordinatorTests
{
    private static (OverlayCoordinator Coordinator, OverlayState State, DrawingStoreFactory Factory) MakeCoordinator()
    {
        var overlayState = new OverlayState();
        var directory = Path.Combine(Path.GetTempPath(), $"OverlayCoordinatorTests-{Guid.NewGuid()}");
        var drawingStoreFactory = new DrawingStoreFactory(new FileDrawingRepository(directory));
        var coordinator = new OverlayCoordinator(
            overlayState,
            drawingStoreFactory,
            new PathStrokeRenderer(),
            new SpyFadeAnimator(),
            new SettingsStore(new InMemorySettingsPersistence()),
            new StubScreenObserving());
        return (coordinator, overlayState, drawingStoreFactory);
    }

    [StaFact]
    public void ToggleNeverMutatesAnyDrawingStore()
    {
        var (coordinator, _, factory) = MakeCoordinator();
        var screen = Screen.PrimaryScreen;
        if (screen is null || DisplayIdentifier.FromScreen(screen) is not { } display) return;

        var store = factory.Store(display);
        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(1, 1));
        store.EndStroke();
        var strokesBeforeToggle = store.Strokes.ToList();

        coordinator.Toggle();
        coordinator.Toggle();

        Assert.Equal(strokesBeforeToggle.Select(s => s.Id), store.Strokes.Select(s => s.Id));
    }

    [StaFact]
    public void ToggleFlipsOverlayVisibility()
    {
        var (coordinator, overlayState, _) = MakeCoordinator();
        Assert.False(overlayState.IsVisible);
        coordinator.Toggle();
        Assert.True(overlayState.IsVisible);
        coordinator.Toggle();
        Assert.False(overlayState.IsVisible);
    }

    [StaFact]
    public void ClearAllNeverMutatesOverlayVisibility()
    {
        var (coordinator, overlayState, _) = MakeCoordinator();
        coordinator.Toggle();
        var visibilityBeforeClear = overlayState.IsVisible;

        coordinator.ClearAll();

        Assert.Equal(visibilityBeforeClear, overlayState.IsVisible);
    }

    [Fact]
    public void ClearAllErasesStrokesOnEveryConnectedDisplay()
    {
        var (coordinator, _, factory) = MakeCoordinator();
        var screen = Screen.PrimaryScreen;
        if (screen is null || DisplayIdentifier.FromScreen(screen) is not { } display) return;

        var store = factory.Store(display);
        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(1, 1));
        store.EndStroke();

        coordinator.ClearAll();

        Assert.Empty(store.Strokes);
    }
}
